### DISCUSSION TITLE
Extending and centering a PNG image removes transparency

### DISCUSSION THREAD

**Discussion by phseiff** [ImageMagick/ImageMagick#3804] [Help] [answered]
## Problem:

I have a PNG image with some transparent areas (they are black underneath the transparency, in case this is relevant).
I am trying to extend and center this image, using

`convert $pic -gravity center -extent "${new_dim}x${new_dim}" $pic`,

where `$pic` is the image and `$new_dim` is the new size (larger than it was before).
This does, however, result in a complete loss of transparency in the image.

## Version:

I am using magick version `6.9.12-14 Q16 x86_64`, in case it is relevant.
I'm also somewhat unsure whether this is a bug or intended behavior, so I am unsure about opening an issue; please tell me in case this'd be the more appropriate channel for this.

## Things I already tried so far:

I tried using `-alpha on`, which was suggested in a thread about the same thing happening with the `-trim` command, but it didn't help.
I also updated magick to the newest 6.x version available, which didn't fix this issue either.
Googling only revealed the same issue to have existed with `-trim` in past versions, yet nothing regarding `-gravity` or `-extent`.

**Comment by fmw42** [comment, marked as answer]
**Works fine for me on IM 6.9.12.14 Q16 Mac OSX Sierra when properly defining the background color.**

The default background is apparently black. So when you leave off -background none, you will get a black background (including where it was transparent). That is the way -extent works by default.

_Create transparent image:_

`convert logo: -transparent white -background black -alpha background logot.png`

![logot](https://user-images.githubusercontent.com/13458218/121114359-6cf03400-c7c8-11eb-9aa9-c94612a8f0f7.png)

**Do extent with default black background:**

`convert logot.png -gravity center -extent 800x800 w.png`

or

`convert logot.png -background black -gravity center -extent 800x800 w.png`

![w](https://user-images.githubusercontent.com/13458218/121116236-0d475800-c7cb-11eb-8af7-ccdd2412e53a.png)



_Do extent with transparent background:_

`convert logot.png -background none -gravity center -extent 800x800 x.png`

![x](https://user-images.githubusercontent.com/13458218/121114399-78dbf600-c7c8-11eb-91e1-4491e8595cdf.png)


**If you want to extend and keep your transparent background but have black outside that, then add -compose copy.**

![y](https://user-images.githubusercontent.com/13458218/121115737-577c0980-c7ca-11eb-8238-955c75d0b0c1.png)

`convert logot.png -gravity center -compose copy -extent 800x800 y.png`

**Reply by phseiff** [reply]
> **Perhaps include -background none as I did above?**

I actually thought I already tried that and it didn't work, but turns out I added `-background` as the last option when it should've been the first one.
Adding it as the first option like you did solved it for me; thanks a lot!

> **What is your output format? If JPEG, then note that JPEG does not support transparency?**
> 
> **What is your platform?**

In case it helps anyone who stumbles upon similar difficulties, my platform is Ubuntu 18.04, and I am invoking Magick via the command line (bash).
My output format was PNG.

EDIT:

Thanks for the additional contextualization, it helps a lot!

**Reply by fmw42** [reply]
Please see my full answer above. I edited it after further investigation.

**Reply by phseiff** [reply]
Thanks, I did :+1:

**Comment by snibgo** [comment]
See the "-extent" documentaton a https://imagemagick.org/script/command-line-options.php#extent :
```
If the image is enlarged, unfilled areas are set to the background color.
```
The default background is opaque white. Use "-background None" to set it to transparent black, and do this _before_ "-extent".

**Reply by fmw42** [reply]
My first reaction was that the default background was white. But the default background for -extent appears to be black not white. See my commands above. The first two produce the same result, that is with or without -background black. So is this a bug in the default background color?

**Reply by snibgo** [reply]
@fmw42: Your first command is:
```
convert logo: -transparent white -background black -alpha background logot.png
```
This stores background=black as metadata in logot.png. So in the next command, which reads that file and uses "-extent" the default background (white) does not apply.

The default applies only if "-background" is not used, and the background has not been read from the image file.

**Reply by fmw42** [reply]
@snibgo. Thanks.  I was not aware that the first command stored the black information for the background in the file.

**Reply by phseiff** [reply]
@snibgo Thanks; that also explains some other minor oddities I stumbled into on the way.
